Support hw-accelerated AES on Thumb and Arm

daverodgman opened the pull request:
## Description

Extend support for AESCE to 32-bit builds.

Saves around 450b for Thumb 2 -Oz (but not for TF-M, because our reference target doesn't have Neon).

Manually tested via qemu, for arm, thumb and aarch64 with gcc and clang.

Do we want to put this in the Changelog as Security, because it enables const-time AES for these targets?

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** no - not a bugfix
- [x] **tests** provided - build test only. Manually tested with qemu

yuhaoth commented on library/aesce.c:
```diff
@@ -64,10 +75,84 @@
+
+#endif /* defined(__clang__) */
+
+static inline uint8x16_t vrbitq_u8(uint8x16_t x)
+{
+    /* There is no vrbitq_u8 instruction in A32/T32, so provide
+     * an equivalent non-Neon implementation. Reverse bit order in each
+     * byte with 4x rbit, rev. */
+    asm ("ldm  %[p], { r2-r5 } \n\t"
+         "rbit r2, r2          \n\t"
+         "rev  r2, r2          \n\t"
+         "rbit r3, r3          \n\t"
+         "rev  r3, r3          \n\t"
+         "rbit r4, r4          \n\t"
+         "rev  r4, r4          \n\t"
+         "rbit r5, r5          \n\t"
+         "rev  r5, r5          \n\t"
+         "stm  %[p], { r2-r5 } \n\t"
+         :
+         /* Output: 16 bytes of memory pointed to by &x */
+         "+m" (*(uint8_t(*)[16]) &x)
+         :
+         [p] "r" (&x)
+         :
+         "r2", "r3", "r4", "r5"
+         );
+    return x;
+}
```
This can reduce 2 instructions. The assembly output is [here](https://godbolt.org/z/vqj7Y1axG)
```suggestion
#include <arm_acle.h>
static inline uint32_t rbit(uint32_t t)
{
    uint32_t ret;
    __asm__("rbit %0,%1":"=r"(ret):"r"(t):);
    return(ret);
}

static inline uint8x16_t vrbitq_u8(uint8x16_t a)
{
    uint32x4_t ret=vreinterpretq_u32_u8(a);
    for(int i=0;i<4;i++)
    {
        ret[i]=__rev(rbit(ret[i]));
    }
    return vreinterpretq_u8_u32(ret);
}
```

tom-cosgrove-arm replied:
Apart from the spacing :) I like this - it's good to have the compiler do as much of the heavy lifting optimisation as possible, and giving it a loop means it's got the chance to unroll or not

yuhaoth replied:
update the spacing. :)

daverodgman replied:
For me, clang -Os generates bigger code (3 more instructions) and less efficient code with this (it's 4x individual load-stores instead of an LDM/STM).

(Also, simpler to use MBEDTLS_BSWAP32 than `__rev`).

yuhaoth replied:
I just update the compilers at [here](https://godbolt.org/z/ccsbK4E6T). GCC and clang 11 are worse than origin code. With `-Os` I got same result.

yuhaoth replied:
And update the suggestion. Replace `__rev`

yuhaoth replied:
Got it.

yuhaoth commented on library/aesce.c:
```diff
@@ -64,10 +75,84 @@
+#if defined(__clang__)
+/* On clang for A32/T32, work around some missing intrinsics and types */
+
+#ifndef vreinterpretq_p64_u8
+#define vreinterpretq_p64_u8 (poly64x2_t)
+#endif
+#ifndef vreinterpretq_u8_p128
+#define vreinterpretq_u8_p128 (uint8x16_t)
+#endif
+#ifndef vreinterpretq_u64_p64
+#define vreinterpretq_u64_p64 (uint64x2_t)
+#endif
```
They are defined as function for aarch64, I do not think it works. And it is bug of clang, According to [ACLE](https://arm-software.github.io/acle/neon_intrinsics/advsimd.html#polynomial-1), A32/T32 mode support those functions.  
```suggestion
#define vreinterpretq_p64_u8 (poly64x2_t)
#define vreinterpretq_u8_p128 (uint8x16_t)
#define vreinterpretq_u64_p64 (uint64x2_t)
```

yuhaoth commented on library/aesce.c:
```diff
@@ -64,10 +75,84 @@
+
+#if defined(MBEDTLS_ARCH_IS_ARM32)
+#if defined(__clang__)
+/* On clang for A32/T32, work around some missing intrinsics and types */
```
```suggestion
/* On clang for A32/T32, work around some missing intrinsics and types which are listed in [ACLE](https://arm-software.github.io/acle/neon_intrinsics/advsimd.html#polynomial-1) */
```

yuhaoth commented on library/aesce.c:
```diff
@@ -64,10 +75,84 @@
 #ifdef __ARM_NEON
 #include <arm_neon.h>
+
+#if defined(MBEDTLS_ARCH_IS_ARM32)
```
```suggestion
#if defined(MBEDTLS_ARCH_IS_ARM32) && defined(MBEDTLS_GCM_C)
```

yuhaoth reviewed (changes requested):
Some comments

yuhaoth commented:
I just push a [commit ](https://github.com/yuhaoth/mbedtls1.3/commit/21ecfe4e08eb73ac44b0f5abb93cf91781ba1c18)to remove plain c code.  `mtest` reports no code size improvement because `mbedtls_internal_aes_[en|de]crypt` use plain c only. 

I test it with below commands. Clang version is 14.0
```
# Plain C only
../mbedtls-docs/tools/bin/mtest -s -MBEDTLS_AESCE_C -MBEDTLS_AES_USE_HARDWARE_ONLY armv8-thumb2
# AESCE only
../mbedtls-docs/tools/bin/mtest -s MBEDTLS_AESCE_C MBEDTLS_AES_USE_HARDWARE_ONLY armv8-thumb2
```
And the result is
| | with commit | Without commit  |
| ------------- |-------------:| -----:|
| Plain C only      | 260446 | 260410 |
| AESCE only     | 258328 |   260652 |
| Result | -2118      |    242 |

yuhaoth commented on library/aesce.c:
```diff
@@ -64,10 +75,84 @@
 #include <arm_neon.h>
+
+#if defined(MBEDTLS_ARCH_IS_ARM32)
+#if defined(__clang__)
```
To fix the CI failure
```suggestion
#if defined(__clang__) && __clang_major__ >= 11
```

yuhaoth replied:
```suggestion
#if defined(__clang__)
#if __clang_major__ < 11
#error "Minimum version of armclang for MBEDTLS_AESCE_C on aarch64 is 6.6."
#endif
```

yuhaoth commented on tests/scripts/all.sh:
```diff
@@ -4354,6 +4354,72 @@ component_build_aes_aesce_armcc () {
+support_build_aes_armce() {
+    # clang >= 4 is required to build with AES extensions
+    ver="$(clang --version|grep version|sed -E 's#.*version ([0-9]+).*#\1#')"
+    [ "${ver}" -ge 4 ]
```
```suggestion
    [ "${ver}" -ge 11 ]
```

yuhaoth reviewed (changes requested):
Just check the clang versions. `poly64*_t` are available from clang 11.0 on aarch32

daverodgman commented:
@yuhaoth I think the existing implementation of `vget_low_p64` for GCC 5 is wrong - looking at `arm_neon.h` different behaviour is needed for big/little endian. I think that the implementation I've done for clang is correct (it is just a trivial wrapper around `vget_low_u64` which already handles endianness) - probably we should use this for gcc as well. WDYT?

daverodgman commented:
> I just push a [commit ](https://github.com/yuhaoth/mbedtls1.3/commit/d974e28c266d42c5854ba21127098ee2460e56e7) to remove plain c code. `mtest` reports no code size improvement because `mbedtls_internal_aes_[en|de]crypt` use plain c .

This looks good - please raise as a separate PR when this is in - let's avoid scope creep on this one.

daverodgman commented:
Removing direct include of `asm/hwcap.h` as it isn't needed (equivalent definitions are pulled in via `sys/auxv.h`) and breaks on our CI (we don't have the Linux headers installed).

daverodgman commented:
@bensze01 please could you add the linux-headers package to the CI (at least on the Ubuntu image) so that we can test the runtime-detection here? thanks

gilles-peskine-arm commented:
> please could you add the linux-headers package to the CI (at least on the Ubuntu image) so that we can test the runtime-detection here? thanks

It's easy to add the package to the docker images, but _should_ we? Compiling a Linux application should not require kernel headers. If we do something that relies on them (and we aren't testing mbedtls as a Linux kernel component), it's good that it fails to compile on the CI: it tells us we're doing something wrong.

daverodgman commented:
> It's easy to add the package to the docker images, but _should_ we? Compiling a Linux application should not require kernel headers. If we do something that relies on them (and we aren't testing mbedtls as a Linux kernel component), it's good that it fails to compile on the CI: it tells us we're doing something wrong.

That is what I thought. On my local Ubuntu 22.04, it's OK to only include `<sys/auxv.h>` because (AIUI) it pulls in a copy of the necessary defines that is intended for user-space to use (in `<bits/hwcap.h>`, provided by `libc6-dev`), so I removed our pre-existing include of `<asm/hwcap.h>` and it worked fine. It didn't work on the CI though - maybe that is lacking `libc6-dev`? So I reverted to including `<asm/hwcap.h>`, but I'm open to suggestions as to the preferred way to do it.

daverodgman commented:
To be clear: the current status in `development` is that we include `<asm/hwcap.h>`, and we don't build-test it in the CI.

gilles-peskine-arm commented:
`sys/auxv.h` is in the [`libc6-dev`](https://packages.ubuntu.com/jammy/libc6-dev) package on on an arm64 Ubuntu. More generally, it's in the glibc dev package on architectures that have this header (it doesn't exist for x86). On other architectures, it's in an extra libc package that's part of the cross-compilation toolchain, e.g. [`linux-libc-dev-armel-cross`](https://packages.ubuntu.com/jammy/linux-libc-dev-armel-cross), which [`libc6-dev-armel-cross`](https://packages.ubuntu.com/jammy/libc-dev-armel-cross). depends on.

I don't know why the cross-compilation toolchain packages are split this way. The native packages have a similar split, but `sys/auxv.h` is in `libc6-dev`, not `linux-libc-dev`.) `linux-libc6-dev-xxx` contains some configuration-agnostic and backward-compatible (but architecture-dependent) kernel headers repackaged by glibc and intended to compile user mode applications, as opposed to `linux-headers-xxx` containing configuration-specific and version-specific kernel headers intended to compile kernel modules. My earlier comment was about the `linux-headers-xxx` packages. I don't see any reason not to install the `linux-libc-dev-xxx` packages.

In fact, we _do_ have the `linux-libc-dev-xxx` packages, because the `libc6-dev-xxx` packages depend on them. But we only have `libc6-dev-armel-cross`, i.e. armv7 without SIMD. What I think we're missing here is arm64 packages, for armv8 support.

yuhaoth commented:
> @yuhaoth I think the existing implementation of `vget_low_p64` for GCC 5 is wrong - looking at `arm_neon.h` different behaviour is needed for big/little endian. I think that the implementation I've done for clang is correct (it is just a trivial wrapper around `vget_low_u64` which already handles endianness) - probably we should use this for gcc as well. WDYT?

Yes, it is wrong.

yuhaoth commented on library/aesce.c:
```diff
@@ -39,19 +48,23 @@
-#   if __clang_major__ < 4
-#       error "Minimum version of Clang for MBEDTLS_AESCE_C is 4.0."
+#   if defined(MBEDTLS_ARCH_IS_ARM32) && (__clang_major__ < 11)
+#       error "Minimum version of Clang for MBEDTLS_AESCE_C on 32-bit Arm or Thumb is 111.0."
```
wrong typo
```suggestion
#       error "Minimum version of Clang for MBEDTLS_AESCE_C on 32-bit Arm or Thumb is 11.0."
```

tom-cosgrove-arm commented on library/sha256.c:
```diff
@@ -180,7 +201,11 @@ static int mbedtls_a64_crypto_sha256_determine_support(void)
+#if defined(MBEDTLS_ARCH_IS_ARM64)
         asm ("sha256h q0, q0, v0.4s" : : : "v0");
+#else
+        asm ("sha256h.32 q0, q0, q0" : : : "q0");
```
Would it make sense to put the `volatile` on these two lines? (I was going to raise a PR to do this, but it would conflict with this one)

daverodgman replied:
DId you mean to make this comment on the SHA PR? I can do that there if you like.

tom-cosgrove-arm replied:
Well, yes, but we should always have `asm volatile`

daverodgman commented:
> `sys/auxv.h` is in the [`libc6-dev`](https://packages.ubuntu.com/jammy/libc6-dev) package on on an arm64 Ubuntu. More generally, it's in the glibc dev package on architectures that have this header (it doesn't exist for x86). On other architectures, it's in an extra libc package that's part of the cross-compilation toolchain, e.g. [`linux-libc-dev-armel-cross`](https://packages.ubuntu.com/jammy/linux-libc-dev-armel-cross), which [`libc6-dev-armel-cross`](https://packages.ubuntu.com/jammy/libc-dev-armel-cross). depends on.
> 
> I don't know why the cross-compilation toolchain packages are split this way. The native packages have a similar split, but `sys/auxv.h` is in `libc6-dev`, not `linux-libc-dev`.) `linux-libc6-dev-xxx` contains some configuration-agnostic and backward-compatible (but architecture-dependent) kernel headers repackaged by glibc and intended to compile user mode applications, as opposed to `linux-headers-xxx` containing configuration-specific and version-specific kernel headers intended to compile kernel modules. My earlier comment was about the `linux-headers-xxx` packages. I don't see any reason not to install the `linux-libc-dev-xxx` packages.
> 
> In fact, we _do_ have the `linux-libc-dev-xxx` packages, because the `libc6-dev-xxx` packages depend on them. But we only have `libc6-dev-armel-cross`, i.e. armv7 without SIMD. What I think we're missing here is arm64 packages, for armv8 support.

Thanks. I think that it's not the arm64 packages that we are missing, but the 32-bit Armv8 packages, i.e. the hard-float ABI packages?

On further investigation on my system (aarch64 Ubuntu 22.04), `bits/hw_cap` provides the aarch64 definitions, but the 32-bit definitions are in `asm/hw_cap`, which is provided by `linux-libc-dev-armhf-cross`. In searching the web, this seems to be how these flags are normally included - see e.g. https://community.arm.com/arm-community-blogs/b/operating-systems-blog/posts/runtime-detection-of-cpu-features-on-an-armv8-a-cpu#:~:text=include%20%3Csys/auxv.h%3E-,%23include%20%3Casm/hwcap.h%3E,-int%20main()%0A%7B%0A%C2%A0%C2%A0%C2%A0%20long%20hwcaps2

So in this case, I think we're missing `linux-libc-dev-armhf-cross` from the CI. And I think the code should include `asm/hwcap.h`.

tom-cosgrove-arm commented on ChangeLog.d/sha256-armce-arm.txt:
```diff
@@ -0,0 +1,4 @@
+Features
+    * Support Armv8 Crypto Extension acceleration for SHA-256
+      when compiling for Thumb or 32-bit Arm.
+
```
(a) extra blank line at the end of the file
(b) PR title references AES, this is about SHA-256 - isn't that #8298?

daverodgman replied:
Again, you've commented on the wrong PR :-)

tom-cosgrove-arm replied:
Is this PR missing `needs-preceding-pr`?

daverodgman replied:
It's based on 8298 in order to pick up the definition of `MBEDTLS_IS_ARMV8`. Yes, I'll add the label.

daverodgman commented:
CI failure is spurious - OpenCI passed

yuhaoth commented:
Remove preceding-PR flag.

yuhaoth reviewed (approved):
LGTM

tom-cosgrove-arm reviewed (approved):
LGTM

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -4391,6 +4391,63 @@ component_build_aes_aesce_armcc () {
+support_build_aes_armce() {
+    # clang >= 4 is required to build with AES extensions
+    ver="$(clang --version|grep version|sed -E 's#.*version ([0-9]+).*#\1#')"
+    [ "${ver}" -ge 11 ]
```
Bah, missed that this doesn't match the comment two lines above. I have a PR touching this function that I'll fix this in

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -4391,6 +4391,63 @@ component_build_aes_aesce_armcc () {
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a32+crypto -mthumb -S"
+    not grep -E 'aes[0-9a-z]+.[0-9]\s*[qv]' library/aesce.o
+    msg "clang, test aarch64 crypto instructions not built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a -S"
```
@tom-cosgrove-arm I'm confused as to exactly in what setting we test the absence of AES instructions. Specifically, why are we testing that on `armv8-a` and not `armv8-a+crypto`? Is that just copypasta or is there a reason?

tom-cosgrove-arm replied:
Actually, I think the errors are on lines 4441 and 4444. Here we are checking that we don't emit AES instructions when we don't ask for them, so that (for example) we won't get SIGILL when run on an older Raspberry Pi (IIRC the RPi5 is the first with crypto extensions). If we passed `+crypto` to the compiler it would be allowed to use any of those instructions itself, so could give false positives. So not having `+crypto` here is correct, but having `+crypto` in the two cases above is incorrect

gilles-peskine-arm replied:
I see, thanks! I'll fix it in https://github.com/Mbed-TLS/mbedtls/pull/9286.

Hmmm...

> we don't emit AES instructions when we don't ask for them

But if we're telling the compiler not to use AES instructions, how would the test ever fail? If we were enabling assembly that calls AES instructions?

gilles-peskine-arm replied:
And below with sha256:

```
            make -B library/sha256.s CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a"
            grep -E 'sha256[a-z0-9]+\s+[qv]' library/sha256.s
```

So we do find sha256 instructions on `armv8-a`, presumably from our assembly?

davidhorstmann-arm replied:
Just following this thread for interest, but I just noticed and was about to raise: does this grepping work? It seems like we are grepping for the assembler mnemonic in the object file, which doesn't seem like it would work, even if we have intrinsics.

Surely we need to `objdump` it first?

tom-cosgrove-arm replied:
The `-S` in `CFLAGS` means "output assembly, not object code, into the output file" (which we still name with`.o`, but hey-ho)

davidhorstmann-arm replied:
Ah okay, makes sense, thanks!

gilles-peskine-arm replied:
```
commit acbdbc7c2b32368b07a5b49a170b202d880afa12
Author: Gilles Peskine <Gilles.Peskine@arm.com>
Date:   2024-06-19 14:16:05 +0200

    Use .s extension for assembly
    
    Having assembly files called *.o was confusing.
```
